kernel/comps/virtio/src/queue.rs

```diff
@@ -32,151 +32,115 @@ pub enum QueueError {
-                            .unwrap(),
-                        true,
-                    )
-                    .unwrap(),
-                    0,
-                ),
-                SafePtr::new(
-                    DmaCoherent::map(
-                        VmAllocOptions::new(1)
-                            .is_contiguous(true)
-                            .alloc_contiguous()
-                            .unwrap(),
-                        true,
-                    )
-                    .unwrap(),
-                    0,
-                ),
-                SafePtr::new(
-                    DmaCoherent::map(
-                        VmAllocOptions::new(1)
-                            .is_contiguous(true)
-                            .alloc_contiguous()
-                            .unwrap(),
-                        true,
-                    )
-                    .unwrap(),
-                    0,
-                ),
-            )
         };
         debug!("queue_desc start paddr:{:x?}", descriptor_ptr.paddr());
         debug!("queue_driver start paddr:{:x?}", avail_ring_ptr.paddr());
         debug!("queue_device start paddr:{:x?}", used_ring_ptr.paddr());
 
+        let queue_size = SIZE as u16;
         transport
-            .set_queue(idx, size, &descriptor_ptr, &avail_ring_ptr, &used_ring_ptr)
+            .set_queue(
+                idx,
+                queue_size,
+                descriptor_ptr.paddr(),
+                avail_ring_ptr.paddr(),
+                used_ring_ptr.paddr(),
+            )
             .unwrap();
-        let mut descs = Vec::with_capacity(size as usize);
-        descs.push(descriptor_ptr);
-        for i in 0..size as usize {
-            let mut desc = descs.get(i).unwrap().clone();
-            desc.add(1);
-            descs.push(desc);
-        }
 
-        let notify = transport.get_notify_ptr(idx).unwrap();
-        // Link descriptors together.
-        for i in 0..(size - 1) {
-            let temp = descs.get(i as usize).unwrap();
-            field_ptr!(temp, Descriptor, next).write(&(i + 1)).unwrap();
-        }
-        field_ptr!(&avail_ring_ptr, AvailRing, flags)
-            .write(&(0u16))
-            .unwrap();
         Ok(VirtQueue {
-            descs,
-            avail: avail_ring_ptr,
-            used: used_ring_ptr,
-            notify,
-            queue_size: size,
-            queue_idx: idx as u32,
-            num_used: 0,
-            free_head: 0,
-            avail_idx: 0,
-            last_used_idx: 0,
+            idx,
+            desc_table: DescTable::new(descriptor_ptr, queue_size),
+            aval_ring: AvailRing::new(avail_ring_ptr),
+            used_ring: UsedRing::new(used_ring_ptr),
+            notify: transport.get_notify_ptr(idx).unwrap(),
         })
     }
 
-    /// Add dma buffers to the virtqueue, return a token.
+    /// Adds dma buffers to the virtqueue, returns a token.
     ///
     /// Ref: linux virtio_ring.c virtqueue_add
```

Comment:
Please document all public methods. The virtio spec has done a great job explaining virtio queues. Read the spec and use the spec's words to explain the operations of virt queue.